WorkloadManagementSystem/Service/JobMonitoringHandler.py

```diff
@@ -63,6 +66,44 @@ def initialize(self):
 
     return S_OK()
 
+  @classmethod
+  def parseSelectors(cls, selectDict=None):
+    """ Parse selectors before DB query
+
+        :param dict selectDict: selectors
+
+        :return: str, str, dict -- start/end date, selectors
+    """
+    selectDict = selectDict or {}
+
+    # Get time period
+    startDate = selectDict.get('FromDate', None)
+    if startDate:
+      del selectDict['FromDate']
+    # For backward compatibility
+    if startDate is None:
+      startDate = selectDict.get('LastUpdate', None)
+      if startDate:
+        del selectDict['LastUpdate']
+    endDate = selectDict.get('ToDate', None)
+    if endDate:
+      del selectDict['ToDate']
+
+    # Provide JobID bound to a specific PilotJobReference
+    # There is no reason to have both PilotJobReference and JobID in selectDict
+    # If that occurs, use the JobID instead of the PilotJobReference
+    pilotJobRefs = selectDict.get('PilotJobReference')
+    if pilotJobRefs:
+      del selectDict['PilotJobReference']
+      if not selectDict.get('JobID'):
+        for pilotJobRef in [pilotJobRefs] if isinstance(pilotJobRefs, six.string_types) else pilotJobRefs:
+          res = cls.pilotManager.getPilotInfo(pilotJobRef)
+          if res['OK'] and 'Jobs' in res['Value'][pilotJobRef]:
+            selectDict['JobID'] = selectDict.get('JobID', [])
+            selectDict['JobID'].extend(res['Value'][pilotJobRef]['Jobs'])
```

Comment:
I guess you don't need to execute this line for each `pilotJobRef` (line 102): 

```
selectDict['JobID'] = selectDict.get('JobID', [])
```
Because:
- If you are here, then it means there is no `JobID` in `selectDict`: (see line 98)
- Once you have it, it is sufficient to extend it, there is no need to get it multiple time.


```suggestion
    if pilotJobRefs:
      del selectDict['PilotJobReference']
      if not selectDict.get('JobID'):
        selectDict['JobID'] = []
        for pilotJobRef in [pilotJobRefs] if isinstance(pilotJobRefs, six.string_types) else pilotJobRefs:
          res = cls.pilotManager.getPilotInfo(pilotJobRef)
          if res['OK'] and 'Jobs' in res['Value'][pilotJobRef]:
            selectDict['JobID'].extend(res['Value'][pilotJobRef]['Jobs'])
```

Reply:
If nothing happens below the 99 line, a `JobID in (  )` query is generated, which also raise a MySQL error.

Reply:
Unless I miss something, `pilotJobRefs` cannot be `None` or empty at this point (see line 96). Am I wrong?

Reply:
It may be, why not?

Reply:
and below the 96th line `pilotJobRefs` can no longer be empty. it's true.